service/raft/group0_voter_registry.cc

```diff
@@ -0,0 +1,198 @@
+                --slots_left_dc;
+                ++itr;
+            }
+
+            // Switch the first non-voters to voters
+            const auto& added_voters = std::ranges::subrange(beg, itr) | std::ranges::views::values;
+            voters_add.insert(added_voters.begin(), added_voters.end());
+
+            const auto& added_voters_as_set = added_voters | std::ranges::views::transform([](const auto& node) {
+                return std::make_pair(true, node);
+            }) | std::ranges::to<std::unordered_multimap>();
+
+            voters.erase(beg, itr);
+            voters.insert(added_voters_as_set.begin(), added_voters_as_set.end());
+        }
+    }
+
+    co_await _voter_client.set_voters_status(voters_add, can_vote::yes, as);
+    co_await _voter_client.set_voters_status(voters_del, can_vote::no, as);
+}
+
+std::unordered_map<sstring, size_t> group0_voter_registry::distribute_slots() {
+    // Calculate the number of voter slots left for each DC
+    std::unordered_map<sstring, size_t> slots_left_per_dc;
+
+    for (const auto& [dc, voters] : _dc_voters) {
+        slots_left_per_dc.emplace(dc, 0);
+    }
+
+    const auto dc_count = _dc_voters.size();
+
+    auto compare_priority_dc = [this](const sstring& dc1, const sstring& dc2) {
+        return _dc_voters[dc2].size() < _dc_voters[dc1].size();
+    };
+
+    std::priority_queue<sstring, std::deque<sstring>, decltype(compare_priority_dc)> dc_by_priority(compare_priority_dc);
+
+    for (const auto& [dc, _] : slots_left_per_dc) {
+        dc_by_priority.push(dc);
+    }
+
+    auto slots_left = std::min(_max_voters, _nodes_count);
+
+    const auto max_slots_per_dc = (dc_count > 2)
+                                          // if the number of DCs is greater than 2, prevent any DC taking majority of voters
+                                          ? (slots_left - 1) / 2
+                                          // with 2 DCs (or less), we can't prevent one DC from taking majority, so we allow more voters per DC
+                                          : slots_left;
+
+    auto dc_count_left = dc_count;
+
+    // Iterate from the DC with the smallest number of nodes to ensure the most even distribution of voters
+    while (!dc_by_priority.empty()) {
+        const auto& dc = dc_by_priority.top();
+
+        auto& slots_left_dc = slots_left_per_dc[dc];
+
+        auto slots_per_dc = slots_left / dc_count_left;
+
+        // Slots for a dc are capped by the number of nodes in the dc
+        slots_left_dc = std::min(slots_per_dc, _dc_voters[dc].size());
+        if (slots_left_dc > max_slots_per_dc) {
+            // If the DC has reached the majority limit, we can't add more
+            rvlogger.debug("Max voters reached for DC: {}, slots: {}", dc, slots_left_dc);
+            slots_left_dc = max_slots_per_dc;
+        }
+        slots_left -= slots_left_dc;
+
+        --dc_count_left;
+        dc_by_priority.pop();
+    }
```

Comment:
Nice implementation of the distribution across dcs. It took me some time to convince myself it works in every case, but it looks like it works indeed.